### ISSUE TITLE
Formatting of section headers

### ISSUE THREAD

**Issue by avekens** [metamath/set.mm#2262]
I have started the revision of graph theory (see issue #1418), and I wonder if there are any (written) conventions how to format section headers and section header comments. I want to format the section headers for graph theory uniformly and according to existing conventions.

Looking at the first headers (e.g. "Propositional calculus"), there should be an empty line between the header and its comment, the comment is indented by two blanks, and there should be an empty line at the end of the comment:

```
$(
#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#
                           Propositional calculus
#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#

  Propositional calculus deals with general truths about well-formed formulas
  (wffs) regardless of how they are constructed.  The simplest propositional
  ...
  for a complete cross-reference from sources used to its formalization in the
  Metamath Proof Explorer.

$)
```

If these are the conventions, we should document them in section 17.1.1 "Conventions" and adjust the headers which do not follow these conventions step by step (I think @benjub is doing this already).

Furthermore, it seems that there should be two empty lines beween the section header and the preceding text:

```
  ${
    idi.1 $e |- ph $.
    $( This inference, which requires no axioms for its proof, is useful as a
       ..
       ~ id .  (Contributed by Alan Sare, 31-Dec-2011.) $)
    idi $p |- ph $=
      (  ) B $.
  $}


$(
#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#
                           Propositional calculus
#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#*#

  Propositional calculus deals with general truths about well-formed formulas
  ...
```
Two questions on that:
* Should this always be the case, even if the preceding text is a section header itself (e.g. between "CLASSICAL FIRST-ORDER LOGIC WITH EQUALITY" and "Pre-logic")? I don't think so.
* Should this be the case for all section levels? I think for the fourth level (-.-.-.-.-.-.-.-.-), this should not be required.

Is there any documentation about the formatting of section headers already available?

**Comment by avekens** [comment]
There was already a discussion about this topic in the Google group: https://groups.google.com/g/metamath/c/1bhQ-e49SCE/m/t7sb8KLlBAAJ

My two questions, however, are not answered by this discussion.

**Comment by benjub** [comment]
Hi @avekens.  Thanks for bringing this topic, and indeed, the foremost rule is consistency.  The spacing used in the prop calc part looks good and readable in my opinion.  I/we should document it in ~conventions (or maybe a new ~conventions-layout or ~conventions-comment or some other label).  The double blank lines before each new section look nice two.  I tried to uniformize both rules from time to time, but not in the whole file.  Then, there are the cases you mention: a section header immediately following a previous one (typically for nested headers), and fourth-level sections.  Maybe, as you say, in the first case double blank lines are not necessary (but in the second case, why treat the fourth level differently from the first three, and why put the limt at the 4th and not the 3rd... the most consistent is to treat everything the same). Again, as much consistency as possible, and as few exceptions as possible.  And rules that make maintenance as light as possible.  Just my two centimes.

**Comment by david-a-wheeler** [comment]
I agree with @benjub - we should document any conventions. For now, I suggest adding them to `~ conventions`, we can always split them out later once they're written down but the bigger challenge is getting them written down :-).

**Comment by benjub** [comment]
I wrote a proposal in PR #2263.  Please, make corrections and/or suggestions there.  Another criterion appeared more obvious to me while writing it: the time and length it takes to write a rule is probably a good indicator of its ease of memorizing, application, and maintenance.

**Comment by nmegill** [comment]
> Looking at the first headers (e.g. "Propositional calculus"), there should be an empty line between the header and its comment, the comment is indented by two blanks, and there should be an empty line at the end of the comment:

When I added the header comment feature, IIRC I started with no blank line
after the header, no indentation, and no blank line at the end of the comment, with the closing "$)" on the same line as the last comment line if it fits.  My reasoning was to allow as much as possible to be displayed in a text editor screen.  Over time people seem to have migrated to the current informal convention that you describe.  If that's what people prefer, it's fine with me.  (Note that `write theorem_list` will produce the same output regardless of indentation and leading and trailing blank lines.)

I have been using a convention of two blank lines before each header, simply because it seemed useful to emphasize visually the separation of sections (like the header decorations themselves, which then became a somewhat arcane "standard" for the computer processing that came later to create the web pages).

Elsewhere, >1 blank lines in a row are changed to 1 line by /rewrap (IIRC; I haven't looked at the code in a while), again to prevent big empty vertical spaces in a limited editor screen that convey little or no information.

 
> Two questions on that:
> 
> * Should this always be the case, even if the preceding text is a section header itself (e.g. between "CLASSICAL FIRST-ORDER LOGIC WITH EQUALITY" and "Pre-logic")? I don't think so.
> * Should this be the case for all section levels? I think for the fourth level (-.-.-.-.-.-.-.-.-), this should not be required.

Personally, for simplicity I would favor a rule of always having two blank lines before each header, regardless of the header level or whether it immediately follows another header.  Then there would be only one rule to remember.  If other people want the above exceptions, though, I won't object.

I don't have plans to modify /rewrap to enforce any double-blank-line convention before the start of headers,

Eventually, I do plan to have /rewrap reformat header comments (right now it leaves them alone), so whatever formatting convention we agree on (like the one you mention) is what it would use.

> 
> Is there any documentation about the formatting of section headers already available?

I don't think so.  There doesn't seem to be a ~conventions section about headers (I found only "Hypertext links to section headers").  `help write theorem_list` provides some basic information about header decorations and levels, which could provide some starting text for whoever would like to write such a section.

**Comment by avekens** [comment]
@nmegill and @benjub thank you for your hints and proposals, I like the introduction part of the conventions in PR #2263. 

Regarding the conventions themselves, I agree with having exceptions as few as possible, i.e. we can demand that "The file set.mm should have a double blank line before section headers, and not at other places.". I also agree with handling all levels uniformly. 

Whether the header comments should be spaced as described (i.e. as those of Part 1) or not, I have no clear opinion. Is it possible to find out who introduced this? As @benjub proposed in PR #2263, we should hear the opinion of others (how can we get them?).  Maybe it would be sufficient to give a hint that lines should not contain unnecessary blanks at the end, because /rewrap would not remove them.

Edit: with "lines should not contain unnecessary blanks at the end" I mean "lines should be filled up as much as possible", because /rewrap would not do this (in contrast to the comments of $a and $p statements). Blanks at the end of a line are not allowed anyway...

**Comment by benjub** [comment]
Another rule that would improve consistency, and that could even be carried out by "WRITE SOURCE [finename] /REWRAP": indentation of section titles (the single lines between the "=-=-=-"-like decorations).  Currently, they are more or less centered, sometimes very inconsistently.  One could settle for an indentation of two white spaces, for instance. What do you think, @nmegill and @avekens ?

**Comment by nmegill** [comment]
> Another rule that would improve consistency, and that could even be carried out by "WRITE SOURCE [finename] /REWRAP": indentation of section titles (the single lines between the "=-=-=-"-like decorations). Currently, they are more or less centered, sometimes very inconsistently. One could settle for an indentation of two white spaces, for instance. What do you think, @nmegill and @avekens ?

I can add this to /rewrap when I add the rewrapping of header comments, although it will be some time before I can get to that.

Originally I attempted to center the titles (by hand) thinking that a centered title would provide a better visual presentation for someone scrolling through the source file, but as you note it has been very inconsistent.  I would be happy with a standard of two spaces of indentation, and if someone wants to do that now (say with the assistance of a quick one-off script) it will be fine with me.

Once /rewrap is updated, it will be just as easy for the code to center the titles as it is to use two spaces of indentation, so it doesn't matter for the long term what we do now.  Converting to 2-space indentation now could give us feedback in case someone prefers the centered approach.

**Comment by avekens** [comment]
> > Another rule that would improve consistency, and that could even be carried out by "WRITE SOURCE [finename] /REWRAP": indentation of section titles (the single lines between the "=-=-=-"-like decorations). Currently, they are more or less centered, sometimes very inconsistently. One could settle for an indentation of two white spaces, for instance. What do you think, @nmegill and @avekens ?
> 
> I can add this to /rewrap when I add the rewrapping of header comments, although it will be some time before I can get to that.
> 
> Originally I attempted to center the titles (by hand) thinking that a centered title would provide a better visual presentation for someone scrolling through the source file, but as you note it has been very inconsistent. I would be happy with a standard of two spaces of indentation, and if someone wants to do that now (say with the assistance of a quick one-off script) it will be fine with me.
> 
> Once /rewrap is updated, it will be just as easy for the code to center the titles as it is to use two spaces of indentation, so it doesn't matter for the long term what we do now. Converting to 2-space indentation now could give us feedback in case someone prefers the centered approach.

I agree with the proposals. I also tried to center the titles (more or less successfully), and I would be happy with a simple rule or automatic formatting. 2-space indentation would be fine.

**Comment by benjub** [comment]
@nmegill : it would be nice to add this to /rewrap indeed.

As for the one-off script, it would be enough to replace all occurrences of the regexp
```
*#*#\n( )^*X
```
with
```
*#*#\n  X
```
where ( )^* stands for "any number of white spaces" and X is any non-white-space character, and similarly with the other section levels (####, -=-=, .-.-). This would probably take script wizards @david-a-wheeler and @jamesjer one minute, but it would take me much more... Any volunteer ?

**Comment by jamesjer** [comment]
It took me longer than one minute and my solution is horribly inefficient.  I'm sure that an awk or sed wizard (I am neither) could do much better.  Create a file named sedexpr.awk and put this in it:
```
#!/bin/awk -f

BEGIN {
    first = 1
    startline = 0
    output = ""
}
END {
    print(output)
}
/^#\*#\*/ {
    if (startline == 0)
	startline = NR + 1
    else {
	if (startline < NR - 1)
	    addr = sprintf("%d,%d", startline, NR - 1)
	else
	    addr = sprintf("%d", startline)
	if (first == 1) {
	    output = sprintf("%ss/^[[:blank:]]*\\(.*\\)/  \\1/", addr)
	    first = 0
	}
	else
	    output = sprintf("%s;%ss/^[[:blank:]]*\\(.*\\)/  \\1/", output, addr)
	startline = 0
    }
}
```

Then invoke it like this (for GNU sed):
```
sed -i "$(awk -f sedexpr.awk set.mm)" set.mm
```
or like this (for BSD sed):
```
sed -i '' "$(awk -f sedexpr.awk set.mm)" set.mm
```

Substitute any other *.mm file for set.mm, but note that the filename appears on the command line twice.

**Comment by benjub** [comment]
It works like a charm ! What is the AWK syntax to match a disjunction of regular expressions ? Namely, I want to replace the above like with something like
```
/^#\*#\* OR ^#### OR ^-=-= OR ^.-.-/ {
```
to match the four header levels.

**Comment by jamesjer** [comment]
Replace the regular expression line with this:
```
/^(#\*#\*|####|-=-=|\.-\.-)/ {
```

**Comment by benjub** [comment]
As for double blank lines preceding each section, the counts in set.mm are, as of today, the following:
```
.\n\n\$\(\n=
359 matches

.\n\n\n\$\(\n=
586 matches

.\n\n\$\(\n-
122 matches

.\n\n\n\$\(\n-
258 matches
```
It's pretty easy to put double blank lines everywhere.  Should I do it ?

**Comment by digama0** [comment]
`write source /rewrap` has something that looks like it is removing occurrences of `\n\n\n` anywhere in the file, so I'm surprised that there are any hits. (EDIT: actually it is `\n\n\n\n`, i.e. 3+ blank lines, that are removed, not 2 blank lines.) I think double blank lines before sections seems fine.

**Comment by benjub** [comment]
@avekens : I think we are done with this issue (#2514 was closed, but replaced by #2522). Can we close it ?

**Comment by avekens** [comment]
Done, mainly by #2514 and #2522.
